subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/worker/TestWorker.java

```diff
@@ -42,18 +42,25 @@
 
 import java.io.Serializable;
 import java.security.AccessControlException;
-import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.LinkedBlockingQueue;
 
 public class TestWorker implements Action<WorkerProcessContext>, RemoteTestClassProcessor, Serializable, Stoppable {
     private static final Logger LOGGER = LoggerFactory.getLogger(TestWorker.class);
     public static final String WORKER_ID_SYS_PROPERTY = "org.gradle.test.worker";
     public static final String WORKER_TMPDIR_SYS_PROPERTY = "org.gradle.internal.worker.tmpdir";
 
     private final WorkerTestClassProcessorFactory factory;
-    private CountDownLatch completed;
+    private LinkedBlockingQueue<TestClassRunInfo> toRun;
     private TestClassProcessor processor;
     private TestResultProcessor resultProcessor;
 
+    private static final TestClassRunInfo SENTINEL_START_STOP = new TestClassRunInfo() {
+        @Override
+        public String getTestClassName() {
+            throw new UnsupportedOperationException();
+        }
+    };
+
```

Comment:
I don't see a problem with the spirit of the change, but the sentinel object feels pretty awkward.  Did you consider implementing this as a wait/notify loop with a state enum?  (happy to elaborate if it's not clear what I mean by that)

Reply:
I did consider a wait/notify loop. That works well enough for marking the `start`, but it doesn't help with marking the end. Maybe something like this?

- wait loop for start (removes need for a start sentinel)
- poll queue for test events
- on stop, do an interrupt rather than send sentinel (removes need for a stop sentinel)

To me, that control flow is harder to read than the existing sentinel logic. If we had Typescript-style union types, I would do `type Event = TestClassRunInfo | "start" | "stop"`, which I think would be the most elegant solution. But since we don't have them, the sentinel seems like the lesser of evils to me, especially since there is only one start and stop and they always happen at the beginning and end.

Happy to revise if I'm misunderstanding or you have a strong preference for a different approach.

Reply:
Good call, I think it's smart to have separate start and stop sentinels, and for those sentinels to have helpful `toString()`. I just pushed up a commit which accomplishes this goal without introducing `interface TestProcessingCommand` and `enum StateChange`, like so:

```java
    private static class StateChangeCommandSentinel implements TestClassRunInfo {
        @Override
        public String getTestClassName() {
            throw new UnsupportedOperationException(toString());
        }

        @Override
        public String toString() {
            String kind = this == START_PROCESSING ? "START" : "STOP";
            return "StateChangeCommandSentinel{" + kind + "_PROCESSING}";
        }
    }

    private static final StateChangeCommandSentinel START_PROCESSING = new StateChangeCommandSentinel();
    private static final StateChangeCommandSentinel STOP_PROCESSING = new StateChangeCommandSentinel();
```

Saves a tiny bit of code and a tiny bit of GC pressure for I think the same level of clarity and debuggability, but I'm happy to copy your demo code verbatim if you prefer.